app/assets/javascripts/admin/subscriptions/controllers/details_controller.js.coffee

```diff
@@ -1,38 +1,43 @@
-angular.module("admin.subscriptions").controller "DetailsController", ($scope, $http, CreditCardResource, StatusMessage) ->
+angular.module("admin.subscriptions").controller "DetailsController", ($scope, $http, CustomerResource, StatusMessage) ->
   $scope.cardRequired = false
 
   $scope.registerNextCallback 'details', ->
     $scope.subscription_form.$submitted = true
-    if $scope.subscription_details_form.$valid
-      $scope.subscription_form.$setPristine()
-      StatusMessage.clear()
-      $scope.setView('address')
-    else
-      StatusMessage.display 'failure', t('admin.subscriptions.details.invalid_error')
+    return unless $scope.validate()
+    $scope.subscription_form.$setPristine()
+    StatusMessage.clear()
+    $scope.setView('address')
 
   $scope.$watch "subscription.customer_id", (newValue, oldValue) ->
     return if !newValue?
-    $scope.loadAddresses(newValue) unless $scope.subscription.id?
-    $scope.loadCreditCards(newValue)
+    $scope.loadCustomer(newValue) unless $scope.subscription.id?
 
   $scope.$watch "subscription.payment_method_id", (newValue, oldValue) ->
     return if !newValue?
     paymentMethod = ($scope.paymentMethods.filter (pm) -> pm.id == newValue)[0]
     return unless paymentMethod?
-    if paymentMethod.type == "Spree::Gateway::StripeConnect"
-      $scope.cardRequired = true
-    else
-      $scope.cardRequired = false
-      $scope.subscription.credit_card_id = null
+    $scope.cardRequired = (paymentMethod.type == "Spree::Gateway::StripeConnect")
```

Comment:
I feel this `type` property is way too coupled to the underlying implementation. Is this Spree imposes?

Reply:
Yes, `Spree::Gateway` is implemented using STI, which I don't like, but we are stuck with it unfortunately.

Reply:
Oh God... STI...

Reply:
Lol, I just googled STI: Sexually Transmissible Infections

But you probably mean Single Table Inheritance.

Reply:
LOL @mkllnk